pkg/objstore/swift/swift.go

```diff
@@ -0,0 +1,267 @@
+package swift
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/improbable-eng/thanos/pkg/objstore"
+
+	"github.com/go-kit/kit/log"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/objectstorage/v1/containers"
+	"github.com/gophercloud/gophercloud/openstack/objectstorage/v1/objects"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const DirDelim = "/"
+
+type swiftConfig struct {
+	AuthUrl       string `yaml:"auth_url"`
+	Username      string `yaml:"username,omitempty"`
+	UserId        string `yaml:"user_id,omitempty"`
+	Password      string `yaml:"password"`
+	DomainId      string `yaml:"domain_id,omitempty"`
+	DomainName    string `yaml:"domain_name,omitempty"`
+	TenantID      string `yaml:"tenant_id,omitempty"`
+	TenantName    string `yaml:"tenant_name,omitempty"`
+	RegionName    string `yaml:"region_name,omitempty"`
+	ContainerName string `yaml:"container_name"`
+}
+
+type Container struct {
+	logger log.Logger
+	client *gophercloud.ServiceClient
+	name   string
+}
+
+func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
+	var sc swiftConfig
+	if err := yaml.Unmarshal(conf, &sc); err != nil {
+		return nil, err
+	}
+
+	authOpts := gophercloud.AuthOptions{
+		IdentityEndpoint: sc.AuthUrl,
+		Username:         sc.Username,
+		UserID:           sc.UserId,
+		Password:         sc.Password,
+		DomainID:         sc.DomainId,
+		DomainName:       sc.DomainName,
+		TenantID:         sc.TenantID,
+		TenantName:       sc.TenantName,
+
+		// Allow Gophercloud to re-authenticate automatically.
+		AllowReauth: true,
+	}
+
+	provider, err := openstack.AuthenticatedClient(authOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := openstack.NewObjectStorageV1(provider, gophercloud.EndpointOpts{
+		Region: sc.RegionName,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Container{
+		logger: logger,
+		client: client,
+		name:   sc.ContainerName,
+	}
+
+	return c, nil
+}
+
+// Name returns the container name for swift.
+func (c *Container) Name() string {
+	return c.name
+}
+
+// Iter calls f for each entry in the given directory. The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (c *Container) Iter(ctx context.Context, dir string, f func(string) error) error {
+	// Ensure the object name actually ends with a dir suffix. Otherwise we'll just iterate the
+	// object itself as one prefix item.
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, DirDelim) + DirDelim
+	}
+
+	options := &objects.ListOpts{Full: false, Prefix: dir, Delimiter: DirDelim}
+	err := objects.List(c.client, c.name, options).EachPage(func(page pagination.Page) (bool, error) {
```

Comment:
Just return this. `err` variable is not needed.

Reply:
done